### DISCUSSION TITLE
mysql-connector-j-8.3.0.jar License Issue

### DISCUSSION THREAD

**Discussion by DL-YYA** [keycloak/keycloak#30484] [Archive]
Hello, everyone.

I'm a newbie to keycloak. I noticed that there is a built-in mysql driver in the distribution: "lib\lib\main\com.mysql.mysql-connector-j-8.3.0.jar", but this driver is licensed under GPL, which is unacceptable for final customer delivery.

I searched all about this but did not find a good solution. So I would like to know if it is possible to let the user decide if this driver needs to be installed or not, like Oracle does now. Or can anyone help with a solution?

Please let me know what you guys think, thanks!

**Comment by jonkoops** [comment]
We're striving to match the Cloud Native Computing Foundation [license policy](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md), any dependencies that do not match the licensing set out there can be considered a bug. 

@stianst is this one of the problematic dependencies you identified, or are we getting an exception for this one?

**Reply by shawkins** [reply]
It does appear that GPLv2 by itself is not allowed.

**Reply by stianst** [reply]
The list there is what licenses are automatically permitted, not about what is disallowed. We have opened requested for exceptions on dependencies including MySQL and waiting for response from the CNCF in this regard.

**Reply by shawkins** [reply]
@stianst thanks for the clarification, that is an important distinction. For reference, here's the link to your exception request - https://github.com/cncf/foundation/issues/817

**Comment by Snehalssg** [comment]
With keycloak 25, do we have any option to ignore mysql driver at build time?

**Reply by shawkins** [reply]
If you are not using the mysql driver, the jar will effectively be ignored: https://github.com/keycloak/keycloak/blob/347f595913d6e309a3d0e986989d4ef617573891/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/IgnoredArtifacts.java

You should also just be able to remove the dependency from the distribution.

**Reply by Snehalssg** [reply]
We have removed the dependency from below location, but it throws error during build time.

Removed jars
/lib/lib/main/com.mysql.mysql-connector-j-8.3.0.jar
/lib/lib/main/io.quarkus.quarkus-jdbc-mysql-3.8.5.jar
/lib/lib/deployment/io.quarkus.quarkus-jdbc-mysql-deployment-3.8.5.jar

After removing jars, if we run build command it throws below error

ERROR: Failed to run 'build' command.
ERROR: java.lang.IllegalArgumentException: D:\Users\AmyData\DSIS\ser\ds_security_server\lib\lib\main\io.quarkus.quarkus-jdbc-mysql-3.8.5.jar does not exist
ERROR: D:\Users\AmyData\DSIS\ser\ds_security_server\lib\lib\main\io.quarkus.quarkus-jdbc-mysql-3.8.5.jar does not exist
ERROR: D:\Users\AmyData\DSIS\ser\ds_security_server\lib\lib\main\io.quarkus.quarkus-jdbc-mysql-3.8.5.jar

We are using postgres db, using keycolak.conf we are mentioning the db details.

**Reply by shawkins** [reply]
You probably don't need to remove the quarkus dependencies.

When you get an exception like this you can try rebuilding the quarkus classloading index  - https://github.com/keycloak/keycloak/blob/a17fd5566962d54b24e54a16f6598c9e36856e62/docs/documentation/server_development/topics/providers.adoc?plain=1#L244

**Reply by lmacko-s1** [reply]
@Snehalssg have you managed to start the keycloak after removal `/lib/lib/main/com.mysql.mysql-connector-j-8.3.0.jar`? it fails for me even with `-Dquarkus.launch.rebuild=true`

**Comment by stianst** [comment]
Could you elaborate on why this is not acceptable? MySQL is not the only GPL (or LGPL) dependency that is bundled with the Keycloak server btw.
